validmind/unit_metrics/__init__.py

```diff
@@ -0,0 +1,192 @@
+import pandas as pd
+import json
+import hashlib
+import importlib
+
+from ..utils import get_model_info
+
+
+global_metric_values = {}
+
+def _serialize_params(params):
+    """
+    Serialize the parameters to a unique hash.
+
+    This function serializes the parameters dictionary to a JSON string,
+    then creates a SHA-256 hash of the string to ensure a unique identifier
+    for the parameters.
+
+    Args:
+        params (dict): The parameters to be serialized.
+
+    Returns:
+        str: A SHA-256 hash of the JSON string representation of the params.
+    """
+    # Convert the params to a JSON string, ensuring consistent ordering
+    params_json = json.dumps(params, sort_keys=True)
+
+    # Create a SHA-256 hash of the JSON string
+    hash_object = hashlib.sha256(params_json.encode())
+    return hash_object.hexdigest()
+
+def _serialize_model(model):
+    """
+    Generate a SHA-256 hash for a scikit-learn model based on its type and parameters.
+
+    Args:
+        model VMModel: The model to be serialized.
+
+    Returns:
+        str: A SHA-256 hash of the model's description.
+    """
+
+    if model is None:
+        # You can return a predefined hash, or handle it in another appropriate way
+        model_info = "none_model_hash"
+    else:
+        model_info = get_model_info(model)
+
+    model_json = json.dumps(model_info, sort_keys=True)
+
+    # Create a SHA-256 hash of the JSON string
+    hash_object = hashlib.sha256(model_json.encode())
+    return hash_object.hexdigest()
+
+def _serialize_dataset(dataset):
```

Comment:
I was doing some testing and this method might end up being too slow and could affect the user experience. I decided to test it with a 4GB dataset and here's what I get:

```
SHA-256 hash: 0626c59781f3c9b5ce224f3d537c65eb301e4c769a35ff9590e0684131192d23
CPU times: user 40.3 s, sys: 2.15 s, total: 42.5 s
Wall time: 42.6 s
```

Since we rely on this to check for cached values, the hash has to be calculated always, meaning that we could end up adding seconds to each metric.

A slight modification could be to implement a "fast hash": we sample a `n` number of rows and calculate their hash. While this is more fragile, we can add the number of rows of the dataset to the calculation so we're a bit protected against change in the number of rows. The sampling could be configured to improve speed but there's a cost to not detecting changes in some rows.

Let's go with a "fast hash" approach for now and we can revisit later. Here's an example implementation:

```python
import pandas as pd
import numpy as np
import hashlib

def fast_hash_with_size(sample_size=1000):
    # Convert the number of rows to bytes and include it in the hash calculation
    rows_bytes = str(len(df)).encode()
    
    # Sample rows if DataFrame is larger than sample_size
    if len(df) > sample_size:
        df_sample = df.sample(n=sample_size, random_state=42)
    else:
        df_sample = df
    
    # Convert the sampled DataFrame to a byte array
    byte_array = df_sample.to_numpy().data.tobytes()
    
    # Initialize the hash object and update it with the row count and data bytes
    hash_obj = hashlib.sha256(rows_bytes + byte_array)
    
    return hash_obj.hexdigest()


# Generate hash
hash_value = fast_hash()
print(f"SHA-256 hash: {hash_value}")
```

Time:

```
SHA-256 hash: f17286b7feb3eeeb3fc6f8d52b49835b6e9148bd7fec02faf39264b4ce54df33
CPU times: user 233 ms, sys: 23.9 ms, total: 257 ms
Wall time: 258 ms
```